project4/name_server/nameserver.cpp

```diff
@@ -1,41 +1,276 @@
+  if (request.id() > id_ && request.id() < successor_id_) {
+    // this entering nameserver's to-be successor will be
+    // this nameserver's successor.
+    request.mutable_successor_info()->set_id(successor_id_);
+    request.mutable_successor_info()->set_port(successor_.port);
+    request.mutable_successor_info()->set_ip(successor_.hostname);
+  } else if (request.id() < id_ && request.id() > predecessor_id_) {
+    // this entering nameserver's predecessor will be
+    // this nameserver's predecessor
+    request.mutable_predecessor_info()->set_port(predecessor_.port);
+    request.mutable_predecessor_info()->set_id(predecessor_id_);
+    request.mutable_predecessor_info()->set_ip(predecessor_.hostname);
+  }
+
+  if (successor_ == bootstrap_ &&
+      !request.mutable_successor_info()->IsInitialized() &&
+      !request.mutable_predecessor_info()->IsInitialized()) {
+    // the entering nameserver will be the last successor in the hash ring.
+    // the entering nameserver's to-be successor is the bootstrap server.
+    // note that the bootstrap server will provide it's predecessor information
+    // (which is this nameserver)
+    request.mutable_successor_info()->set_id(successor_id_);
+    request.mutable_successor_info()->set_port(successor_.port);
+    request.mutable_successor_info()->set_ip(successor_.hostname);
+  }
+  // if the the request does not have predecessor info but has successor info by
+  // the time it reaches back to bootstrap, we know the entering nameserver will
+  // be the last in the ring. if the request does not have successor info but
+  // has predecessor info by the time it reaches back to bootstrap, we know the
+  // entering nameserver will be the first in the ring.
+}
 
 void Nameserver::HandleRequest(
-    const consistent_hash_msgs::ExitInformation& request) {}
+    const consistent_hash_msgs::ExitInformation& request) {
+  // take over the exiting server's key-vals
+  bounds_.second = request.upper_bounds();
+  for (int i = 0; i < request.keys_size(); i++) {
+    std::pair<int, std::string> pair(request.keys(i), request.values(i));
+    pairs_.insert(pair);
+  }
+
+  // update predecessor info
+  predecessor_.port = request.predecessor_info().port();
+  predecessor_.hostname = request.predecessor_info().ip();
+  predecessor_id_ = request.predecessor_info().id();
+}
 
 void Nameserver::HandleRequest(
-    const consistent_hash_msgs::UpdatePredecessorRequest& request) {}
+    consistent_hash_msgs::UpdatePredecessorRequest& request,
+    message_passing::Endpoint source) {
+  // update predecessor info
+  predecessor_.port = request.predecessor_info().port();
+  predecessor_.hostname = source.hostname;
+  predecessor_id_ = request.predecessor_info().id();
+
+  // give lower half of it's key-val range to the entering nameserver
+  consistent_hash_msgs::UpdatePredecessorResponse res;
+  res.set_lower_bounds(bounds_.first);
+  res.set_upper_bounds(bounds_.second / 2);
```

Comment:
Is this supposed to be `res.set_upper_bounds(bounds_.first + (bounds_.second - bounds.first) / 2)` It might be more readable to do something like `const uint kMiddleKey = bounds_.first + (bounds_.second - bounds_.first) / 2;` and then use variable.

Reply:
yup. thanks again.